### DISCUSSION TITLE
Using a Moose Action Instead of [Mesh] in Input

### DISCUSSION THREAD

**Discussion by Vandenbg** [idaholab/moose#26012] [Q&A General]
Just as the title suggests, I am developing an action which will replace the majority of my input deck. The last piece I would like to include is the [Mesh] block. I do not understand C++ that well and I have not worked with Actions before, the majority of my work has been copy/ paste and variations. Is there a good example somewhere of how I can generate a mesh within my action?

**Comment by lindsayad** [comment]
Not that I'm aware of, although you can look at the action that we use to create the mesh which is `SetupMeshAction`

**Reply by Vandenbg** [reply]
Okay, thank you, I'll start there.

**Reply by lindsayad** [reply]
And then also `SetupMeshCompleteAction` ... but that action will get auto-built and will always execute no matter what you do in your input file

**Reply by GiudGiud** [reply]
See some example code in THM maybe.
THM simulations regularly do NOT have a mesh block. 

 @joshuahansel do you know where this is done?

**Reply by joshuahansel** [reply]
THM might not be a good example because it has a lot of other stuff going on, and it's making its own mesh and such. I'll add a different answer.

**Comment by joshuahansel** [comment]
If your mesh is something that you could just make in a `Mesh` block, I think you want to register your action to the task `meta_action`. Check out `TestAction::act()` (this lives in the thermal_hydraulics module). It essentially hides
```
[Mesh]
  type = GeneratedMesh
  dim = 1
  nx = somenumber
[]
```
within the action. The magic happens in `TestAction::addMeshInternal`.

Basically, it's just creating the actions that would be created if you put the above in your input file, and then adding those to the action warehouse (really early).
